src/mlpack/methods/decision_tree/decision_tree_regressor.hpp

```diff
@@ -0,0 +1,527 @@
+                   std::remove_reference<WeightsType>::type>::value>* = 0);
+
+  /**
+   * Make prediction for the given point, using the entire tree.  The predicted
+   * label is returned.
+   *
+   * @param point Point to predict.
+   */
+  template<typename VecType>
+  double Predict(const VecType& point) const;
+
+  /**
+   * Make prediction for the given points, using the entire tree. The predicted
+   * labels for each point are stored in the given vector.
+   *
+   * @param data Set of points to predict.
+   * @param predictions This will be filled with predictions for each point.
+   */
+  template<typename MatType>
+  void Predict(const MatType& data,
+                arma::Row<double>& predictions) const;
+
+  /**
+   * Serialize the tree.
+   */
+  template<typename Archive>
+  void serialize(Archive& ar, const uint32_t /* version */);
+
+  //! Get the number of children.
+  size_t NumChildren() const { return children.size(); }
+
+  //! Get the number of leaves in the tree.
+  size_t NumLeaves() const;
+
+  //! Get the child of the given index.
+  const DecisionTreeRegressor& Child(const size_t i) const { return *children[i]; }
+  //! Modify the child of the given index (be careful!).
+  DecisionTreeRegressor& Child(const size_t i) { return *children[i]; }
+
+  //! Get the split dimension (only meaningful if this is a non-leaf in a
+  //! trained tree).
+  size_t SplitDimension() const { return splitDimension; }
+
+  /**
+   * Given a point and that this node is not a leaf, calculate the index of the
+   * child node this point would go towards.  This method is primarily used by
+   * the Predict() function, but it can be used in a standalone sense too.
+   *
+   * @param point Point to predict.
+   */
+  template<typename VecType>
+  size_t CalculateDirection(const VecType& point) const;
+
+ private:
+  //! The vector of children.
+  std::vector<DecisionTreeRegressor*> children;
+  //! The dimension this node splits on.
+  size_t splitDimension;
+  //! The type of the dimension that we have split on (only meaningful if this
+  //! is a non-leaf in a trained tree).
+  size_t dimensionType;
+  /**
+   * This variable may hold different things. If the node has no children, then
+   * it is guaranteed to hold the prediction label for that node.  If the node
+   * has children, then it may be used arbitrarily by the split type's
+   * CalculateDirection() and SplitIfBetter() function. In this case, it stores
+   * the point at which the split was made.
+   */
+  double splitPointOrPrediction;
+
+  //! Note that this class will also hold the members of the NumericSplit and
+  //! CategoricalSplit AuxiliarySplitInfo classes, since it inherits from them.
+  //! We'll define some convenience typedefs here.
+  typedef typename NumericSplit::AuxiliarySplitInfo
+      NumericAuxiliarySplitInfo;
+  typedef typename CategoricalSplit::AuxiliarySplitInfo
+      CategoricalAuxiliarySplitInfo;
```

Comment:
After reading the other comments as well, I think this is an important issue which we should address. Honestly, I have no knowledge about memory management. So, whatever you just said went over my head. I can take whatever you suggest as the best approach as a grain of salt and can implement that. :)

Reply:
Sure, basically what I think at this point might work best is if:

 - `DecisionTreeClassifier` holds a `std::vector<double>` instead of an `arma::vec` for `classProbabilities`
 - `DecisionTreeRegressor` should hold a `double` for the prediction, just like it does now (so no change needed)
 - `AuxiliarySplitInfo` for `BestBinaryNumericSplit` should be a `double` that holds the split value, and `classProbabilities` shouldn't be used
 - `AuxiliarySplitInfo` for `AllCategoricalSplit` should just be a `size_t` holding the number of categories, and `classProbabilities` shouldn't be used

If you want, I'm happy to spend some more time explaining what's guiding my intuition above.  If you like reading, Ulrich Drepper's [What Every Programmer Should Know About Memory](https://people.freebsd.org/~lstewart/articles/cpumemory.pdf) is an *awesome* article (though it is quite long!) that I learned a *lot* from.  I can give you the short summary too, just let me know. :+1:

Reply:
Thanks for the detailed strategy. The starting part where Ulrich is explaining the memory architecture seems familiar. This semester, I had Computer Organisation course where these architectures were taught on hardware level. But, yeah the article is too big to read right now, if you could give me some highlights then it would be great. :)

Reply:
The "shortest version" is just that sequential memory accesses are comparatively cheap, and random memory accesses are not.  So if you can "stream" through data, it will be fast, but if you have to jump around in memory, it probably won't be.  If I have a vector in memory `[1, 2, 3, 4, 5, 6, ..., N]`, summing all the elements in that vector will be very fast since I can do that as a simple single pass through contiguous memory.  Hardware is often tuned so that if you request the memory address `M`, it is very fast to load the address `M + c` where `c` is, e.g., the cache line size or something similar.  On the other hand, if I represented that vector as a linked list, then every element could live somewhere completely different in memory---and hardware tends to be not as good at this.

The "second shortest version" adds the observations that chunks of data are loaded into cache at a time, and once you have something in the L1/L2/L3 cache, it is very fast to operate on!

So, while keeping in mind that all intuition needs to be validated with experiments, it is generally a decent rule of thumb for design to try and make memory access as linear as possible, and try to make all the memory you need fit into as few cache lines as possible.  For a decision tree this is hard, since the tree structure itself means we won't always have a linear traversal in memory.  But, we can at least try and make sure that all of the decision tree fits into as few cache lines as possible.

Anyway, if you find this kind of thing interesting, keep reading and keep paying attention in the computer architecture classes! :smile: There is some *really cool* stuff to know about all this (at least in my opinion :)).

Reply:
Thanks for the tip. I will keep it in mind. Knowing about these architectures and how they work gives programmers good intuition to write efficient code and get the most out of the processor. I too find it interesting.